File: .claude/skills/sast-scan/rules/semgrep/java/tests/TestMiscSecurity.java
import java.io.*;
import java.net.URL;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Random;
import javax.el.ExpressionFactory;
import javax.naming.directory.DirContext;
import javax.net.ssl.X509TrustManager;
import javax.xml.parsers.DocumentBuilderFactory;
import org.springframework.web.multipart.MultipartFile;

public class TestMiscSecurity extends HttpServlet {
    private String sharedState = "unsafe";

    public void vulnerable(
        String userInput,
        HttpServletRequest request,
        HttpServletResponse response,
        MultipartFile file,
        DirContext ctx,
        ObjectInputStream ois,
        TemplateEngine tmpl,
        VelocityEngine vel,
        DriverManager dm
    ) throws Exception {
        // ruleid: java.security.ssrf-url-open
        new URL(userInput).openConnection();

        // ok: java.security.ssrf-url-open
        new URL("https://api.example.com").openConnection();

        // ruleid: java.security.path-traversal-file
        new FileInputStream(userInput);

        // ok: java.security.path-traversal-file
        new File("/etc/hosts");

        // ruleid: java.security.xxe-xml-parser
        DocumentBuilderFactory.newInstance();

        // ok: java.security.xxe-xml-parser
        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);

        // ruleid: java.security.insecure-random
        new Random();

        // ok: java.security.insecure-random
        new SecureRandom();

        // ruleid: java.security.spring-csrf-disabled
        http.csrf().disable();

        // ruleid: java.security.spring-cors-wildcard
        registry.allowedOrigins("*");

        // ruleid: java.security.spring-permit-all
        auth.permitAll();

        // ruleid: java.security.weak-hash-md5
        MessageDigest.getInstance("MD5");

        // ruleid: java.security.weak-hash-sha1
        MessageDigest.getInstance("SHA-1");

        // ok: java.security.weak-hash-sha1
        MessageDigest.getInstance("SHA-256");

        class TrustAllManager implements X509TrustManager {
            // ruleid: java.security.insecure-ssl-trustall
            public void checkClientTrusted(java.security.cert.X509Certificate[] chain, String authType) { }
            // ruleid: java.security.insecure-ssl-trustall
            public void checkServerTrusted(java.security.cert.X509Certificate[] chain, String authType) { }
            // ruleid: java.security.insecure-ssl-trustall
            public java.security.cert.X509Certificate[] getAcceptedIssuers() { return new java.security.cert.X509Certificate[0]; }
        }

        // ruleid: java.security.deserialization-objectinput
        ois.readObject();

        // ruleid: java.security.el-injection
        fac.createValueExpression(userInput, Object.class);

        // ok: java.security.el-injection
        fac.createValueExpression("${safe}", Object.class);

        // ruleid: java.security.template-injection
        tmpl.process(userInput, ctx);

        // ok: java.security.template-injection
        tmpl.process("safe-template", ctx);

        // ruleid: java.security.ldap-injection
        ctx.search("ou=users", "(uid=" + userInput + ")", null);

        // ok: java.security.ldap-injection
        ctx.search("ou=users", "(uid=alice)", null);

        // ruleid: java.security.log-injection
        log.info(request.getParameter("username"));

        // ok: java.security.log-injection
        log.info("user login");

        // ruleid: java.security.multipart-upload
        request.getFile("upload");

        // ok: java.security.multipart-upload
        if (file.getSize() > 1024) { return; }
        request.getFile("upload");

        // ruleid: java.security.hardcoded-jdbc-password
        DriverManager.getConnection(url, "user", "pass123");

        // ruleid: java.security.thread-safety-servlet
        sharedState = userInput;

        // ruleid: java.security.tempfile-insecure
        File.createTempFile("report", ".tmp");

        // ruleid: java.security.runtime-exec-shell
        new ProcessBuilder("sh", "-c", userInput);
    }
}

File: .claude/skills/sast-scan/rules/semgrep/go/tests/test_misc_security.go
package main

import (
	"bytes"
	cryptorand "crypto/rand"
	"crypto/md5"
	"crypto/sha1"
	"crypto/tls"
	"encoding/gob"
	"fmt"
	"html/template"
	"log"
	mathrand "math/rand"
	"net/http"
	"os"
	"path/filepath"
)

type User struct{}
type fileInfo struct{ Size int64 }
type gormStub struct{}

func (g gormStub) Where(query string, args ...any) gormStub { return g }
func (g gormStub) Find(dest any)                            {}

var (
	w      http.ResponseWriter
	info   fileInfo
	cache  = map[string]string{}
	gormDB gormStub
)

func miscVulnerable(client *http.Client, req *http.Request, db any, url string, base string, userInput string, data []byte) {
	// ruleid: go.security.defer-close
	resp, err := http.Get("https://api.example.com")
	_ = resp
	_ = err

	// ok: go.security.defer-close
	safeResp, safeErr := http.Get("https://api.example.com")
	if safeErr == nil {
		defer safeResp.Body.Close()
	}

	// ruleid: go.security.ssrf-http-get
	http.Get(url)

	// ruleid: go.security.ssrf-http-client
	client.Get(url)

	// ok: go.security.ssrf-http-client
	client.Get("https://api.example.com")

	// ruleid: go.security.path-traversal
	os.Open(filepath.Join(base, userInput))

	// ok: go.security.path-traversal
	os.Open(filepath.Join("/srv/data", "static.txt"))

	// ruleid: go.security.insecure-random
	mathrand.Intn(10)

	// ok: go.security.insecure-random
	cryptorand.Read(data)

	// ruleid: go.security.template-injection
	template.Must(template.New(userInput))

	// ok: go.security.template-injection
	template.Must(template.New("fixed-template"))

	// ruleid: go.security.log-injection
	log.Printf(userInput)

	// ok: go.security.log-injection
	log.Printf("user=%s", userInput)

	// ruleid: go.security.header-injection
	w.Header().Set("X-User", userInput)

	// ok: go.security.header-injection
	w.Header().Set("X-User", "alice")

	// ruleid: go.security.multipart-upload
	req.FormFile("upload")

	// ok: go.security.multipart-upload
	if info.Size > 1024 {
		_, _ = req.FormFile("upload")
	}

	// ruleid: go.security.weak-hash-md5
	md5.Sum(data)

	// ruleid: go.security.weak-hash-sha1
	sha1.New()

	// ruleid: go.security.insecure-tls-config
	tls.Config{InsecureSkipVerify: true}

	// ruleid: go.security.deserialization-gob
	gob.NewDecoder(bytes.NewReader(data)).Decode(&User{})

	go func() {
		// ruleid: go.security.race-condition-map
		cache[userInput] = "value"
	}()

	// ruleid: go.security.hardcoded-credentials
	password := "supersecret123"
	_ = password

	// ok: go.security.hardcoded-credentials
	password = os.Getenv("DB_PASSWORD")

	// ruleid: go.security.tempfile-insecure
	os.CreateTemp("/tmp", "report")

	// ruleid: go.security.file-permission
	os.OpenFile("data.txt", os.O_CREATE, 0666)

	// ruleid: go.security.env-default-secret
	apiSecret := os.Getenv("SECRET")
	if apiSecret == "" {
		apiSecret = "fallback-secret"
	}

	// ok: go.security.env-default-secret
	configuredSecret := os.Getenv("SECRET")
	if configuredSecret == "" {
		return
	}

	// ruleid: go.security.sql-injection-gorm
	gormDB.Where(fmt.Sprintf("id = %s", userInput)).Find(&User{})

	// ruleid: go.security.goroutine-leak
	go func() {
		log.Println("running")
	}()
}
